### ISSUE TITLE
[DMP 2025]: Enabling Safer Roads Through AI and Blockchain - A Scalable Government Solution

### ISSUE THREAD

**Issue by seetadev** [seetadev/RoadICM#1]
## Description

RoadICM is a next-generation road incident management platform that combines Artificial Intelligence with blockchain technology to enhance road safety and emergency response capabilities. The system leverages AI-based object detection, GPS tracking, and video analytics integrated with the Near blockchain and IPFS/Filecoin for secure, real-time incident monitoring and data storage. It is designed to operate efficiently with minimal hardware—such as standard mobile phones and basic camera setups—making it a highly scalable and cost-effective solution for governments at all levels.

**Why It Is Needed:**  
Traditional road monitoring systems are fragmented, reactive, and often lack real-time data, leading to delays in response and risk mitigation. There is also a lack of transparency and secure data sharing between citizens and authorities. RoadICM addresses these gaps by providing a proactive, automated, and decentralized solution that ensures tamper-proof data recording, improved coordination, and citizen involvement in road safety initiatives.

**What It Will Accomplish:**

- Enable **real-time detection** of collisions, abnormal driving behavior, and road hazards.
- Provide **secure, immutable records** of incidents using blockchain for accountability and transparency.
- Facilitate **faster emergency response** with immediate access to footage and metadata.
- Promote **citizen participation** through decentralized incident reporting and feedback tools.
- Support **data-driven policy making** by aggregating environmental, behavioral, and road condition data.
- Offer a **cost-effective, easily deployable** platform using existing digital infrastructure.

## Goals

- [ ] **Implement AI-Powered Object Detection System**  
  Deploy real-time object and behavior detection using CNN-based models to identify accidents, road hazards, and abnormal driver activity via basic camera setups.

- [ ] **Integrate Secure, Tamper-Proof Blockchain Data Logging**  
  Utilize Near Protocol and IPFS/Filecoin to securely store video snapshots, metadata, GPS data, and citizen-generated reports with full data integrity and auditability.

- [ ] **Enable Citizen Reporting and Feedback Tools**  
  Launch a decentralized mobile/web platform for community participation—citizens can report issues, submit footage, and receive real-time updates and resolution status.

- [ ] **Establish Smart Dashboard and Live Monitoring Tools**  
  Create a centralized dashboard for government authorities to view live feeds, alarms, air quality data (via OpenAQ), and vehicle diagnostics in real time.

### ✅ **Goals to be Achieved By Mid-point Milestone**

- [x] Functional integration of Near blockchain + IPFS for secure incident recording  
- [x] Live object detection demo deployed using AWS, PyTorch/TensorFlow models  
- [x] NFT-based digital ID system for licenses and vehicle registrations created  
- [x] Feedback collection tool using NEAR Book module implemented  
- [x] MVP website and streaming dashboard shared publicly

## **Key Performance Indicators (KPIs)**

- 🎯 90%+ incident detection accuracy for real-time object recognition  
- ⏱️ <10 seconds latency for event-to-alert pipeline  
- 🛡️ 100% of stored data linked to verifiable blockchain hashes  
- 📈 Onboard 500+ citizen users for pilot feedback participation  
- 📊 Generate analytics reports for 10+ incident types using smart dashboard

## **Government Collaboration Milestones**

- 🏛️ **Pilot Program Proposal to be Submitted** to Ministry of Road Transport & Highways  
- 🧪 **Initial Field Deployment** to be planned with Smart City/Transport Authority in 1 urban zone (NCT, Delhi). 
- 🧾 **MOU to be Signed** for Data Sharing & Coordination  
- 🚦 **Interoperability Tests** with Existing Traffic and CCTV Systems  
- 📢 **Joint Public Awareness Campaign** to encourage citizen usage of the reporting platform

## **Setup / Installation Guide**

Below is a step-by-step guide to set up and run the **RoadICM** system components:

### 🔧 **Frontend Setup (React + UI5 + HTML5)**

1. Clone the repository:  
   ```bash
   git clone https://github.com/seetadev/RoadICM.git
   cd RoadICM/near road ICM/Near-analytics-tool
   ```

2. Install dependencies:  
   ```bash
   npm install
   ```

3. Start the development server:  
   ```bash
   npm run dev
   ```

   - 📍 Vercel Deployment Preview: [View Live](https://aspiring-road-safety-near-xrpl-deploy.vercel.app/)
   - 🌐 Developer Tools: [NEAR Web3 Tooling](https://web3-tools-near-ecosystem-deploy.vercel.app/)

### 🤖 **Object Detection (AI/ML Model)**

- **Tools Used:** TensorFlow / PyTorch / OpenCV
- Follow the object detection setup [here](https://github.com/seetadev/RoadICM/tree/main/video-analytics-pipeline)
  1. Install dependencies from `requirements.txt`
  2. Connect your camera or video feed source
  3. Run the detection pipeline:
     ```bash
     python detect.py --source=0  # or use video file
     ```

### 🌐 **Blockchain Integration (Near Protocol)**

1. Install NEAR CLI:  
   ```bash
   npm install -g near-cli
   ```

2. Login to NEAR testnet:  
   ```bash
   near login
   ```

3. Deploy a smart contract (NFT or Feedback system):  
   - Example repo: [NFT Vehicle Registration](https://github.com/seetadev/RoadICM/tree/main/near%20road%20ICM/Near-Vehicle-Registration-NFT)

   - Compile & deploy using:
     ```bash
     npm run build
     near deploy --accountId <your-account.testnet> --wasmFile out/main.wasm
     ```

---

### 📦 **IPFS + [NFT.Storage](https://nft.storage/) Integration**

- Store snapshots and metadata:
  - Go to [NFT.Storage](https://nft.storage/)
  - Upload your image or data
  - Use returned IPFS hash and link it with Near smart contract

---

### 📡 **Live Streaming + Alarm Management**

- Use AWS/open source WebRTC toolkits for video streaming  
- Demo setup: [Live Video Dashboard + Alarm](https://aspiring-road-safety-near-xrpl-deploy.vercel.app/)
- Alarm configuration linked to smart contract + NFT marketplace (Sequence wallet + Turtle Music)

### Expected Outcome

The final **RoadICM** system will be a robust, scalable, and decentralized **Road Incident and Safety Management Platform** for use by **government agencies, city administrations, and citizens**. The system will seamlessly combine AI-powered video analytics, blockchain-backed data integrity, and citizen engagement into a single, easy-to-deploy platform.

### 🔍 **System Behavior and Capabilities**

1. **Real-Time Object and Incident Detection**
   - Detect road anomalies, collisions, and abnormal driver behavior using live video feeds from dashcams or mobile devices.
   - Instant alerts pushed to authorities and displayed on smart dashboards.

2. **Secure Incident Logging**
   - Automatically store video snapshots, alarm metadata, and reports on IPFS/Filecoin.
   - Log verifiable hashes and metadata on the NEAR blockchain for tamper-proof evidence.

3. **Citizen Feedback and Reporting Portal**
   - Users can submit incident reports, upload videos/images, and provide feedback on traffic or road conditions via a mobile/web portal.
   - Submissions are stored transparently on-chain, accessible to relevant authorities.

4. **Vehicle and Driver Credential Management**
   - NFT-based registration of driver licenses and vehicle ownership.
   - Blockchain-powered validation using NEAR Protocol standards (NEP-171).

5. **Smart Government Dashboard**
   - Real-time visual display of road incidents, air quality (via OpenAQ), and fleet health.
   - Searchable logs, filters for incident type/location/time, and downloadable reports.

6. **Live Video Streaming + Alarm Marketplace**
   - Streaming of ongoing road activities with object detection overlays.
   - Custom alarm tones/wallpapers purchasable via NFT marketplace (Sequence + Turtle Music integration).

7. **Fleet & Road Asset Monitoring**
   - Monitor blackbox data, GPS routes, vehicle diagnostics.
   - Track usage and availability of road surveillance assets using NFC-tagged devices with NFT records.

### 🎯 **High-Level Outcomes**

- 📉 **Reduction in emergency response time** due to real-time alerts and location-aware incident tracking.
- 🔐 **Increased transparency and accountability** in road safety data handling.
- 🧠 **Data-driven policy and planning** with secure analytics from both citizens and automated systems.
- 👥 **Stronger public engagement and trust** through participatory safety reporting.
- 🛠️ **Lower deployment costs** by using mobile-first, serverless, open-source technologies.

### 🔍 **1. Real-Time Detection & Alerting**
- [ ] System detects and classifies objects (vehicles, pedestrians, roadblocks, etc.) with ≥90% accuracy in live video streams.
- [ ] Real-time alerts are generated within 10 seconds of incident detection.
- [ ] Alerts are logged and visible in the government dashboard with timestamp and location metadata.

### 🛡️ **2. Secure and Verifiable Data Storage**
- [ ] All incident video snapshots are uploaded to IPFS/Filecoin using NFT.Storage.
- [ ] Corresponding metadata and IPFS hashes are immutably recorded on NEAR blockchain.
- [ ] Each record can be independently verified via blockchain explorer.

### 👥 **3. Citizen Feedback & Reporting**
- [ ] Citizens can log in using NEAR wallet and submit feedback/reports through a public web portal.
- [ ] Submitted reports must include optional media files and geolocation tags.
- [ ] Government administrators can respond to reports and update their status (pending/resolved/archived).

### 🪪 **4. NFT-Based Licensing and Registration**
- [ ] Users can mint NFTs for their driving licenses and vehicle IDs.
- [ ] The system validates user ID (SSN) and vehicle ID (VRN) ranges correctly (0–65535).
- [ ] Smart contract follows NEP-171 standards and passes unit tests on NEAR testnet.

### 📊 **5. Dashboard Functionality**
- [ ] Government dashboard displays:
   - Live streaming feeds with detection overlays.
   - Real-time alarm notifications.
   - Filters by incident type, location, and time.
   - Downloadable reports and analytics.
- [ ] Air quality data is integrated via OpenAQ API and displayed in real-time.

### 🔄 **6. Interoperability and Extensibility**
- [ ] The system can integrate with third-party CCTV systems via API or streaming source URL.
- [ ] Supports plug-in modules for AI models and blockchain networks.
- [ ] NFC-enabled monitoring devices can be onboarded and tracked through the NFT marketplace.

### 📱 **7. Mobile-Friendly & Serverless**
- [ ] The platform works on standard mobile devices using browser or app interface.
- [ ] All storage and computation rely on decentralized services (Fluence, IPFS, NEAR).
- [ ] No centralized server is required to operate core functionality.

### Implementation Details

The **RoadICM** system is built on a decentralized architecture combining AI, blockchain, and IoT tools. Below are the core components and their implementation specifics:

### 🧠 **1. AI-Based Object Detection**

- **Tech Stack:** Python, OpenCV, TensorFlow / PyTorch  
- **Implementation:**
  - Real-time object detection pipeline using CNN models.
  - Integration with camera feeds (dashcams, smartphones).
  - Detection output (bounding boxes, class labels) streamed or logged.
- **Deployment:** Edge-based detection on mobile, browser, or embedded hardware.

### 🔗 **2. Blockchain & Decentralized Storage**

- **Blockchain Protocol:** [[NEAR Protocol](https://near.org/)](https://near.org)
- **Data Storage:** [[NFT.Storage](https://nft.storage/)](https://nft.storage/) for IPFS/Filecoin
- **Use Cases:**
  - Smart contract for incident logging, vehicle/license NFT registration.
  - Immutable logs for video snapshots and metadata.
  - Citizen feedback and audit trail of actions by government officials.

### 📦 **3. Smart Contracts**

- **Languages:** Rust (for NEAR), AssemblyScript (where required)
- **Standards:** NEP-171 (NFT standard), NEAR CLI for contract management
- **Modules:**
  - Vehicle and License NFT Contracts
  - Feedback Book for Incident Reports
  - NFT Marketplace for surveillance assets and alarm personalization

### 📺 **4. Frontend Interface**

- **Frameworks:** React.js, SAP UI5, HTML5  
- **Deployment:** [[Vercel](https://vercel.com/)](https://vercel.com/)  
- **Features:**
  - Live dashboard for incident monitoring
  - Alarm and object detection overlay
  - Citizen reporting interface with NEAR wallet integration

### 🛰️ **5. Streaming and Alarm Management**

- **Video Streaming:** WebRTC / AWS Kinesis Video Streams (optional)
- **Alarm Storage:**
  - Snapshot + metadata saved to IPFS
  - IPFS CID linked to alarm data on-chain
- **Alarm Viewer:** Custom UI to view, set, or delete alarms, integrated with NFT-backed customization (e.g., wallpapers, ringtones)

### 🛰️ **6. Data Analytics & Air Quality**

- **Data Sources:** [[ASDI OpenAQ](https://openaq.org/)](https://openaq.org/)
- **Tooling:** NEAR low-code analytics tools (custom dashboard)
- **Usage:**
  - Fetch air quality data by location
  - Display pollution index on live dashboard
  - Inform decisions for collision detection + driver alerting

### 🔐 **7. Wallets, Identity & DAO Tooling**

- **Wallets:** NEAR Wallet, Sequence Wallet (for music/NFT auction features)
- **DAO Integration:** DAO tooling for coordination among ministries, transport bodies, and community safety groups.
- **User Identity:** Mapped via 16-bit SSN & VRN integer IDs; each user associated with a NEAR address.

### 🌍 **8. External Integrations**

- **NFC Devices:** NFC tags assigned to CCTV, drones, and other road-side monitoring equipment.
- **Govt Asset Marketplace:** NFT-based registry for state surveillance tools and road assets.
- **Interoperability:** API endpoints for integration with Smart City dashboards and traffic police databases.

### Mockups/Wireframes

Please visit: https://drive.google.com/drive/u/3/folders/10hXK3s2y1or1oA2HKJThPZLn27xkNOQp

### Product Name

Near Transport Intel

### Organisation Name

NSUT in collaboration with C4GT, SEETA and AIC

### Domain

Open Source Library

### Tech Skills Needed

Artificial Intelligence, Chart.js, AWS, DevOps, Machine Learning, Ionic, TypeScript

### Mentor(s)

@seetadev @aspiringsecurity @aspiringdevapps

Manu Sheel Gupta;

Dr. MPS Bhatia;

Deepti Gupta (SAP Appgyer guidance and SAP BTP integrations) for utility government organizations in India.

Category

### Category

Breaking Change, Beginner Friendly, Machine Learning, Research, Analytics, Backend

**Comment by dikshant182004** [comment]
hii @seetadev @aspiringsecurity @aspiringdevapps ,i would like to contribute in this project and enable safer roads for my society ,I have one question, since i am a Machine learning enthusiasts ,so making things related to AI would not be a big deal for me .However i don't have much knowledge in making smart contracts or blockchain .so do i need to have that skill beforehand for applying in this project or i can learn it during the course of our project and then apply it .

**Comment by jprabhat** [comment]
Hi @seetadev @aspiringsecurity @aspiringdevapps , 
I'm really excited about this project and would love to contribute, particularly to the Web3-related components.
I'm interested in contributing to the following areas:

🔐 Wallets, Identity & DAO Tooling
=> Integration of wallets like NEAR Wallet and Sequence Wallet (especially for music/NFT auction features).
=> DAO tooling for coordination between ministries, transport bodies, and community safety groups.
=> Decentralized user identity mapping via 16-bit SSN & VRN integer IDs, linked to NEAR addresses.

🛡️ Secure and Verifiable Data Storage
=> Uploading video snapshots to IPFS/Filecoin using NFT.Storage.
=> Immutably recording metadata and IPFS hashes on the NEAR blockchain.
=> Supporting verifiability via blockchain explorers.

I’d be happy to help implement and improve these features. Looking forward to contributing and learning with the community!

**Comment by NandhiniN0506** [comment]
Hello @seetadev,

I came across the National Vehicle Maintenance & Mobility Infrastructure System using Web3 Technologies initiative, and I’m genuinely excited about its vision and scope. The integration of decentralized technologies like Sign Protocol, Filecoin IPC, Arbitrum Stylus, Neon EVM, and Risc0 ZKVM to digitize and streamline vehicle maintenance, monitoring, and incentivization represents an incredibly meaningful leap forward for India’s transport infrastructure.

I’m highly interested in contributing to this project—particularly in the areas of verifiable recordkeeping, decentralized monitoring, and smart incentivization systems. The opportunity to build scalable, privacy-preserving, and interoperable systems for such a high-impact domain aligns perfectly with my aspirations and skill development goals. I’m also eager to collaborate with the mentors and the broader open-source community to help realize this ambitious vision.

Please let me know how I can get involved!

Best regards,
Nandhini

**Comment by 29Ra7jn8iSu0th0ar** [comment]
Hello Sir,
I am interested in this project. I want to work more on skills Artificial Intelligence, Chart.js, AWS, DevOps, Machine Learning, Ionic, TypeScript.so I kindly request you to consider my application.

warm regards, 
Rajni Suthar

**Comment by yashh13kumar** [comment]
Hi @seetadev @aspiringsecurity @aspiringdevapps,
I'm highly interested in contributing to this project. I believe I can effectively implement the proposed system by leveraging computer vision for real-time incident detection, blockchain (NEAR + IPFS) for tamper-proof data logging, and multi-agent AI to automate verification, escalation, and reporting workflows. Excited about the opportunity to build a scalable, decentralized solution for road safety!

**Comment by sravika1914** [comment]
Contribution Proposal: AI-Based Object Detection Integration + Blockchain Learning Commitment

Hi @seetadev 

I’m excited to contribute to RoadICM and support its vision for intelligent, decentralized road safety. Below is a summary of the contribution I plan to make:

Proposed Contribution Scope

1. AI-Powered Object Detection

Implement and fine-tune CNN-based models using PyTorch/TensorFlow for detecting:

  * Accidents (collisions, rollovers)
  * Road hazards (debris, potholes)
  * Abnormal driver behavior (wrong-way driving, sudden stops)
* Use OpenCV for real-time video feed integration and frame extraction.
* Prepare detection overlay outputs for downstream streaming and alerting.

2. Integration with RoadICM Pipeline

* Format detection output for blockchain compatibility (i.e., snapshot and metadata for IPFS upload).
* Enable test run/demo using webcam or sample dashcam footage.

3. Learning and Growing in Blockchain

* I’m actively learning NEAR Protocol and smart contract development.
* I plan to work on simple Rust-based contracts and understand NFT.Storage and IPFS workflows.
* Will gradually extend my contribution to areas like:

  * NFT-based driver/vehicle IDs
  * Tamper-proof incident logs
  * Blockchain-linked alarm notifications

Tools & Stack

* ML/AI: Python, PyTorch, OpenCV
* Blockchain (learning): NEAR CLI, Rust, IPFS, NFT.Storage
* Frontend (optional extension): React, Vercel deployment testing

Goals

* Deliver a working object detection demo (video-to-detection-to-IPFS snapshot)
* Collaborate on blockchain integration as I ramp up my skills
* Contribute actively and learn alongside the community

Looking forward to feedback, suggestions, and mentorship as I begin contributing.

— Sravika Kadali

**Comment by irfu-git** [comment]
This project aligns perfectly with my background in AWS Cloud, DevOps, and networking. I'm excited to contribute scalable, secure infrastructure and automation to support real-time road safety solutions.

**Comment by Mrkbhavan** [comment]
I m interested to contribute to this project and have excited to join in open source DMP 2025.
